File: CSharpCodeBase/aiactions/sleep.lua
local Class = require 'utils.hump.class'
local AINode = require 'utils.ailib.ainode'
local AISequenceNode = require 'utils.ailib.aisequence'

local AI_Sleep = Class({__includes=AINode, init = function (self, entity, sleepPhases)
  AINode.init(self, entity)
  self.sleepPhases = sleepPhases or {GameController.daytime.PHASE_EVENING, GameController.daytime.PHASE_NIGHT}
end})

function AI_Sleep:CheckSleepPhase()
  return Tables.Find(self.sleepPhases, GameController.daytime.currentPhase) > -1
end

function AI_Sleep:Visit()
  if self.status == NODE_READY then
    if self:CheckSleepPhase() and self.entity:HasEmptyHitlist() and math.chance(75) then
      self.status = NODE_RUNNING
      self.entity:Sleep()
      self.sleepScanTimeout = self.entity.config:Get('sleepScanTimeout') or 5
    else
      self.status = NODE_FAILURE
      self.parent:Sleep(10)
    end
  end
  
  if self.status == NODE_RUNNING then
    self.sleepScanTimeout = self.sleepScanTimeout - GameController.deltaTime
    if self.sleepScanTimeout < 0 then
      self.sleepScanTimeout = self.entity.config:Get('sleepScanTimeout') or 5
      local targets = self.entity.vision:GetVisibleEntitiesByRSTag("enemy", self.entity.config:Get("visionRadius") * 0.3)
      if #targets > 0 then
        if math.chance(self.entity.config:Get('sleepStopChance') or 10) then
          self.status = NODE_SUCCESS
          self.entity:WakeUp()
          self.entity:ScanEnemies()
        end
      end
    end
    if not self:CheckSleepPhase() or not self.entity.isSleeping then
      self.status = NODE_SUCCESS
      --self.entity:WakeUp()
    end
  end
  return self.status
end

local AI_WakeUp = Class({__includes=AINode, init = function (self, entity)
  AINode.init(self, entity)
end})

function AI_WakeUp:Visit()
  if self.status == NODE_READY then
    if self.entity.isSleeping then
      self.entity:WakeUp()
      self.status = NODE_RUNNING
    else
      self.status = NODE_FAILURE
    end
  end
  
  if self.status == NODE_RUNNING then
    if self.entity.mecanim:CheckStateName('Idle') then
      print("Awakened!")
      self.status = NODE_SUCCESS
    end
  end
  return self.status
end

local AI_SleepCycle = Class({__includes=AISequenceNode, init = function (self, entity, sleepPhases)
      local childNodes = {
        AI_Sleep(entity, sleepPhases),
        AI_WakeUp(entity),
      }
      AISequenceNode.init(self, entity, childNodes)
end})


return AI_SleepCycle

File: CSharpCodeBase/aiactions/sleep.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class sleep :ainode{
 var AI_Sleep = Class({__includes=AINode, init = public void (self, entity, sleepPhases){
   AINode.init(self, entity);
   this.sleepPhases = sleepPhases  ||  {GameController.daytime.PHASE_EVENING, GameController.daytime.PHASE_NIGHT}
 }})
 public void AI_CheckSleepPhase(){
   return Tables.Find(this.sleepPhases, GameController.daytime.currentPhase) > -1;
 }
 public void AI_Visit(){
   if(this.status == NODE_READY  ){ 
     if(self:CheckSleepPhase()  &&  this.entity:HasEmptyHitlist()  &&  math.chance(75)  ){ 
       this.status = NODE_RUNNING;
       this.entity:Sleep();
       this.sleepScanTimeout = this.entity.config:Get("sleepScanTimeout")  ||  5;
     }else{
       this.status = NODE_FAILURE;
       this.parent:Sleep(10);
     }
   }
   
   if(this.status == NODE_RUNNING  ){ 
     this.sleepScanTimeout = this.sleepScanTimeout - GameController.deltaTime;
     if(this.sleepScanTimeout < 0  ){ 
       this.sleepScanTimeout = this.entity.config:Get("sleepScanTimeout")  ||  5;
       var targets = this.entity.vision:GetVisibleEntitiesByRSTag("enemy", this.entity.config:Get("visionRadius") * 0.3);
       if(#targets > 0  ){ 
         if(math.chance(this.entity.config:Get("sleepStopChance")  ||  10)  ){ 
           this.status = NODE_SUCCESS;
           this.entity:WakeUp();
           this.entity:ScanEnemies();
         }
       }
     }
     if(not self:CheckSleepPhase()  ||  not this.entity.isSleeping  ){ 
       this.status = NODE_SUCCESS;
       //this.entity:WakeUp()
     }
   }
   return this.status;
 }
 var AI_WakeUp = Class({__includes=AINode, init = public void (self, entity){
   AINode.init(self, entity);
 }})
 public void AI_WakeUp:Visit(){
   if(this.status == NODE_READY  ){ 
     if(this.entity.isSleeping  ){ 
       this.entity:WakeUp();
       this.status = NODE_RUNNING;
     }else{
       this.status = NODE_FAILURE;
     }
   }
   
   if(this.status == NODE_RUNNING  ){ 
     if(this.entity.mecanim:CheckStateName("Idle")  ){ 
       print("Awakened!");
       this.status = NODE_SUCCESS;
     }
   }
   return this.status;
 }
 var AI_SleepCycle = Class({__includes=AISequenceNode, init = public void (self, entity, sleepPhases){
       var childNodes = {
         AI_Sleep(entity, sleepPhases),;
         AI_WakeUp(entity),;
       }
       AISequenceNode.init(self, entity, childNodes);
 }})
 return AI_SleepCycle}}